File: Bookcase core/src/main/java/cz/vhromada/book/facade/AuthorFacade.java
package cz.vhromada.book.facade;

import cz.vhromada.book.entity.Author;
import cz.vhromada.common.facade.MovableParentFacade;
import cz.vhromada.result.Result;

/**
 * An interface represents facade for authors.
 *
 * @author Vladimir Hromada
 */
public interface AuthorFacade extends MovableParentFacade<Author> {

    /**
     * Adds author. Sets new ID and position.
     * <br>
     * Validation errors:
     * <ul>
     * <li>Author is null</li>
     * <li>ID isn't null</li>
     * <li>Position isn't null</li>
     * <li>First name is null</li>
     * <li>First name is empty string</li>
     * <li>Middle name is empty string</li>
     * <li>Last name is null</li>
     * <li>Last name is empty string</li>
     * </ul>
     *
     * @param data author
     * @return result with validation errors
     */
    @Override
    Result<Void> add(Author data);

    /**
     * Updates author.
     * <br>
     * Validation errors:
     * <ul>
     * <li>Author is null</li>
     * <li>ID is null</li>
     * <li>Position is null</li>
     * <li>First name is null</li>
     * <li>First name is empty string</li>
     * <li>Middle name is empty string</li>
     * <li>Last name is null</li>
     * <li>Last name is empty string</li>
     * <li>Author doesn't exist in data storage</li>
     * </ul>
     *
     * @param data new value of author
     * @return result with validation errors
     */
    @Override
    Result<Void> update(Author data);

}

File: Bookcase core/src/main/kotlin/cz/vhromada/book/facade/AuthorFacade.kt
package cz.vhromada.book.facade

import cz.vhromada.book.entity.Author
import cz.vhromada.result.Result

/**
 * An interface represents facade for authors.
 *
 * @author Vladimir Hromada
 */
interface AuthorFacade : BookcaseFacade<Author> {

    /**
     * Adds data. Sets new ID and position.
     *
     * Validation errors:
     * * ID isn't null
     * * First name is empty string
     * * Middle name is empty string
     * * Last name is empty string
     *
     * @param data data
     * @return result with validation errors
     */
    override fun add(data: Author): Result<Void>

    /**
     * Updates data.
     *
     * Validation errors:
     * * ID is null
     * * First name is empty string
     * * Middle name is empty string
     * * Last name is empty string
     * * Author doesn't exist in data storage
     *
     * @param data new value of data
     * @return result with validation errors
     */
    override fun update(data: Author): Result<Void>

}
